client/wfm/src/service/WfmService.ts

```diff
@@ -0,0 +1,248 @@
+    protected resultService: ResultService,
+    protected userService: UserService
+  ) {
+  }
+
+  /**
+   * Beginning a workflow for a single Workorder.
+   *
+   * @param {string} workorderId - The ID of the workorder to begin the workflow for.
+   */
+  public beginWorkflow(workorderId): Promise<StepOperationData> {
+    return this.workorderSummary(workorderId).then(summary => {
+      if (summary.result) {
+        console.warn(`beginWorkflow() called on already started workflow with id: ${workorderId},`
+          + ` ignoring existing result`);
+      }
+      const { workorder, workflow, nextStepIndex } = summary;
+      return this.createNewResult(workorderId, workorder.assignee)
+        .then(result => {
+          // Now we check the current status of the workflow to see where the next step should be.
+          result.status = this.checkStatus(workorder, workflow, result);
+
+          // We now have the current status of the workflow for this workorder, the begin step is now complete.
+          return {
+            workorder,
+            workflow,
+            result,
+            nextStepIndex,
+            step: nextStepIndex > -1 ? workflow.steps[nextStepIndex] : workflow.steps[0]
+          };
+        });
+    });
+  }
+
+  /**
+   * Gets a summary of the workflow.
+   * This will get all of the details related to the workorder, including workflow and result data.
+   *
+   * @param {string} workorderId - The ID of the workorder to get the summary for.
+   * @return {{workflow: Workflow, workorder: Workorder, result: Result}}
+   * An object containing all the major entities related to the workorder
+   */
+  public workorderSummary(workorderId): Promise<Summary> {
+    return this.workorderService.read(workorderId)
+      .then(workorder => Promise.all([
+        this.workflowService.read(workorder.workflowId),
+        this.resultService.readByWorkorder(workorderId)
+      ]).then(([workflow, result]) => ({
+        workflow,
+        workorder,
+        result,
+        nextStepIndex: this.executeStepReview(workflow.steps, result).nextStepIndex
+      })));
+  }
+
+  /**
+   *
+   * Going to the previous step of a workorder.
+   *
+   * @param workorderId - The ID of the workorder to switch to the previous step for
+   */
+  public previousStep(workorderId): Promise<StepOperationData> {
+    const self = this;
+    return this.workorderSummary(workorderId).then(summary => {
+      const { workorder, workflow, result } = summary;
+      let { nextStepIndex } = summary;
+
+      if (!result) {
+        // No result exists, The workflow should have been started
+        return Promise.reject(new Error('No result exists for workflow ' + workorderId +
+          '. The workflow back topic can only be used for a workflow that has begun'));
+      }
+
+      // -1 is a special value for 'no next step'
+      nextStepIndex = _.max([nextStepIndex - 1, -1]) || -1; // _.max returns `number?`
```

Comment:
I think that nextStepIndex here is crucial point, that may be problematic to do things we plan in future. 
If we code it so it will be state machine. I would really like nextStepIndex to disappear and apply rule decision facade that will change current step. Doing this math operations doesn't contribute to visiblity and flexibility.
Imagine that someone want to add some conditions to decide where to go. 

I think I have idea how to make it flexible and clean. Let's agree on something and possibly create new ticket. 
I will be able to migrate from `nextStepIndex` fairly fast by adding some simple `StepTransitionEngine` that can be extended in future (for example to meet BPMn spec)